src/mapping/query/function/parse_url.rs

```diff
@@ -0,0 +1,129 @@
+use super::prelude::*;
+use std::collections::BTreeMap;
+use std::iter::FromIterator;
+use url::Url;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct ParseUrlFn {
+    query: Box<dyn Function>,
+}
+
+impl ParseUrlFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for ParseUrlFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let bytes = required!(ctx, self.query, Value::Bytes(v) => v);
+
+        Url::parse(&String::from_utf8_lossy(&bytes))
+            .map_err(|e| format!("unable to parse url: {}", e))
+            .map(Into::into)
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Bytes(_)),
+            required: true,
+        }]
+    }
+}
+
+impl TryFrom<ArgumentList> for ParseUrlFn {
+    type Error = String;
+
+    fn try_from(mut arguments: ArgumentList) -> Result<Self> {
+        let query = arguments.required("value")?;
+
+        Ok(Self { query })
+    }
+}
+
+impl From<Url> for Value {
+    fn from(url: Url) -> Self {
+        let mut map = BTreeMap::<&str, Value>::new();
+
+        map.insert("scheme", url.scheme().to_owned().into());
+        map.insert("username", url.username().to_owned().into());
```

Comment:
`username` is always set, but will be an empty string if not defined, whereas `password` can be `null`. This is based on the [`url` crate](https://docs.rs/url) (which is based off https://url.spec.whatwg.org/).

Although, having said that, the spec itself states that both the username and password are an empty string initially, whereas the `url` crate explicitly makes the former an empty string, and the latter an `Option`, so if anyone thinks we should change this, I'm happy to do so.

Reply:
It would seem to make sense to make these two consistent. I would have picked defaulting to `null` instead of an empty string, but if the spec says differently then that is probably preferable.

Reply:
I found https://github.com/servo/rust-url/issues/234, which describes how the spec changed after the crate was already released.

I assume they haven't updated the signature to prevent backward breaking changes.

I'll go ahead and make both an empty string by default.